File: PDOO/Practica4/Java/Irrgarten/src/irrgarten/Shield.java
/**
 * @file Shield.java
 * @author Miguel Angel De la Vega Rodríguez
 * @author Alberto De la Vera Sanchez
 * @brief This file contains the class that represents the shield which 
 * can be used by the player in the game to protect himself from a monster
 */

package irrgarten;

public class Shield extends CombatElement{
    
    /**
     * Constructor for Shield class, receives protection value and amount of uses
     * 
     * @param protection Protection provided by the shield
     * @param uses Uses of the shield
     */
    public Shield(float protection, int uses) {
        super(protection, uses);
    }

    /**
     * Returns the protection value of the shield if it has uses left
     * and decreases the number of uses by one and returns 0 otherwise,
     * delegating in the method produceEffect
     * 
     * @return the protection value of the shield
     */
    public float protect() {
        return this.produceEffect();
    }

    /**
     * Returns a string with the current state of the shield in the following
     * format:
     * 
     * S[Protection, Uses]
     * 
     * @return a string with the current state of the shield
     */
    @Override
    public String toString() {
        return "S" + super.toString();
    }

}

File: PDOO/Practica5/Ruby/Irrgarten/shield.rb
require_relative 'dice'
require_relative 'combat_element'

module Irrgarten
  class Shield < CombatElement
    # Returns the protection value of the shield if it has uses left, or 0.0 otherwise
    # @return [Float] The protection value of the shield
    def protect
      produce_effect
    end

    # Returns whether the shield is discarded or not based on the number of uses left
    # @return [Boolean] whether the shield is discarded or not
    def discard
      Dice.discard_element(@uses)
    end

    # Returns a string representation of the shield
    # @return [String] The string representation of the shield
    def to_s
      "S#{super}"
    end
  end
end
